private/bufpkg/bufconfig/module_config.go

```diff
@@ -78,6 +80,19 @@ type ModuleConfig interface {
 	//
 	// This may be nil.
 	ModuleFullName() bufmodule.ModuleFullName
+	// Includes says the module only includes these paths. The paths are relative to DirPath.
```

Comment:
The assymetry between `Includes` being a `[]string` and `RootToExcludes` including roots and being a `map[string][]string`, with the excludes in the map being root-relative and the includes in the slice being dir-path-relative (is there a difference?) is confusing, and isn't explained in the documentation. As a user/reader of this API, I don't immediately understand, which is the burden of proof going forward. The documentation should address:

- Why includes are relative to a different directory than excludes, if they are.
- Why there is the difference in signature between includes and excludes.

I bet the answer will be some form of "includes are v2 only", which I don't think is a good answer here - these configs are meant to be version-agnostic as much as possible. I think you'll likely end up changing the signature to `RootToIncludes` to match the excludes signature.

Reply:
The disconnect here from my perspective is that we are using `map[string][]string` to handle the `v1beta1` keys, `roots` and `excludes` on a single `ModuleConfig`. As the documentation for `RootsToExcludes` states, we should only ever expect a single root for `v1` and `v2` configs:

https://github.com/bufbuild/buf/blob/8063ca9cced372fb8c288121083b107f88540cd3/private/bufpkg/bufconfig/module_config.go#L105

This is because a `Module` should only ever have one root and multiple roots are handled as a `ModuleSet` -- but `v1beta1` breaks that concept, because the `roots` key in the `v1beta1` configuration is more akin to defining a `ModuleSet` rather than a single `Module`. The proper fix to have it so that a single `ModuleConfig` and `Module` has only one root and still supporting `v1beta1` configurations should be that `v1beta1` configs would have one `ModuleConfig` per root, and then at the targeting layer only return a single `ModuleReadBucket` for all of these configs, since they should be built and used as one module (and they should be resolved as a single `OpaqueID`).

However, this feels a little bit out of scope for this PR, and maybe should be a fast follow-up, but I am not sure that we should lean into the fact that at the `ModuleConfig` layer, we have a construct that supports multiple roots for a single `Module` when that is not the intended structure on configurations going forward.

It is also worth noting that `v1beta1` enforces rules on its `roots` that we are breaking in in `v2`:

https://github.com/bufbuild/buf/blob/8063ca9cced372fb8c288121083b107f88540cd3/private/bufpkg/bufconfig/module_config.go#L85-L88

Reply:
I don't understand most of the above to be honest, perhaps need to talk offline. This, however:

> It is also worth noting that v1beta1 enforces rules on its roots that we are breaking in v2:

This is not clear to me from our discussions/this PR/the changelog/etc that it is now intended to allow `foo/bar` and `foo`. Why is this required? I was just about to do QA to make sure this had not changed, and if it needs to, we need to talk about it.